### PR TITLE
[WIP] Eager agg

### PR DESCRIPTION
**Pull request by iyerr3**
Query:
```
SELECT min(s1)
FROM foo, bar
WHERE foo.j1 = bar.j2
AND   foo.s1 % 2 = 0
GROUP BY foo.g1, bar.g2
```

Plan without EagerAgg
```
                                                  QUERY PLAN
--------------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice2; segments: 3)  (cost=0.00..43788.53 rows=562500 width=4)
   ->  Result  (cost=0.00..43780.14 rows=187500 width=4)
         ->  HashAggregate  (cost=0.00..43780.14 rows=187500 width=4)
               Group Key: foo.g1, bar.g2
               ->  Redistribute Motion 3:3  (slice1; segments: 3)  (cost=0.00..43734.55 rows=187500 width=12)
                     Hash Key: foo.g1, bar.g2
                     ->  Result  (cost=0.00..43727.51 rows=187500 width=12)
                           ->  HashAggregate  (cost=0.00..43727.51 rows=187500 width=12)
                                 Group Key: foo.g1, bar.g2
                                 ->  Hash Join  (cost=0.00..11032.52 rows=132874743 width=12)
                                       Hash Cond: (foo.j1 = bar.j2)
                                       ->  Table Scan on foo  (cost=0.00..2595.31 rows=13333572 width=12)
                                             Filter: ((s1 % 2) = 0)
                                       ->  Hash  (cost=431.77..431.77 rows=33341 width=8)
                                             ->  Table Scan on bar  (cost=0.00..431.77 rows=33341 width=8)
 Optimizer: PQO version 3.1.0
```

Plan with EagerAgg
```
                                                   QUERY PLAN
----------------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice2; segments: 3)  (cost=0.00..12090.26 rows=562500 width=4)
   ->  Result  (cost=0.00..12081.88 rows=187500 width=4)
         ->  HashAggregate  (cost=0.00..12081.88 rows=187500 width=4)
               Group Key: foo.g1, bar.g2
               ->  Redistribute Motion 3:3  (slice1; segments: 3)  (cost=0.00..12036.29 rows=187500 width=12)
                     Hash Key: foo.g1, bar.g2
                     ->  Result  (cost=0.00..12029.25 rows=187500 width=12)
                           ->  HashAggregate  (cost=0.00..12029.25 rows=187500 width=12)
                                 Group Key: foo.g1, bar.g2
                                 ->  Hash Join  (cost=0.00..7414.04 rows=18754313 width=12)
                                       Hash Cond: (foo.j1 = bar.j2)
                                       ->  HashAggregate  (cost=0.00..5843.87 rows=1881938 width=12)
                                             Group Key: foo.j1, foo.g1
                                             ->  Table Scan on foo  (cost=0.00..2595.31 rows=13333572 width=12)
                                                   Filter: ((s1 % 2) = 0)
                                       ->  Hash  (cost=431.77..431.77 rows=33341 width=8)
                                             ->  Table Scan on bar  (cost=0.00..431.77 rows=33341 width=8)
 Optimizer: PQO version 3.1.0
```
14bd814

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@iyerr3 Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/pivotal?repositoryId=greenplum-db/gporca&pullRequestId=412)!

[Click here](https://cla.pivotal.io/sync/pivotal?repositoryId=greenplum-db/gporca&pullRequestId=412) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@iyerr3 Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/pivotal?repositoryId=greenplum-db/gporca&pullRequestId=412)!

File: libgpopt/include/gpopt/xforms/CXformEagerAgg.h [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+	//---------------------------------------------------------------------------
+	class CXformEagerAgg : public CXformExploration
+	{
+
+		private:
+
+			// private copy ctor
+			CXformEagerAgg(const CXformEagerAgg &);
+        
+
```
Comment (vraghavan78) [inline libgpopt/include/gpopt/xforms/CXformEagerAgg.h]: indentation is screwed up

File: libgpopt/include/gpopt/xforms/CXformEagerAgg.h [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+             IMemoryPool *mp, // memory pool
+             CExpression *orig_proj_list,       // project list of the original global aggregate
+             CExpression **lower_proj_list, // project list of the new local aggregate
+             CExpression **upper_proj_list // project list of the new global aggregate
+            );
+
+            // return true if xform should be applied only once
+            virtual
+            BOOL IsApplyOnce(){
+                return true;
```
Comment (vraghavan78) [inline libgpopt/include/gpopt/xforms/CXformEagerAgg.h]: can u add why

File: libgpopt/src/xforms/CXformEagerAgg.cpp [hunk 3]
```diff
@@ -0,0 +1,390 @@
+//---------------------------------------------------------------------------i
+//  Greenplum Database
+//  Copyright (C) 2018 Pivotal Inc.
+//
+//  @filename:
+//      CXformEagerAgg.cpp
+//
+//  @doc:
+//      Implementation for eagerly pushing aggregates below join
+//          (with on restriction on the join condition)
```
Comment (vraghavan78) [inline libgpopt/src/xforms/CXformEagerAgg.cpp]: can u rephrase?

File: libgpopt/src/xforms/CXformEagerAgg.cpp [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+using namespace gpmd;
+
+//---------------------------------------------------------------------------
+//  @function:
+//      CXformEagerAgg::CXformEagerAgg
+//
+//  @doc:
+//      Ctor
+//
+//---------------------------------------------------------------------------
```
Comment (vraghavan78) [inline libgpopt/src/xforms/CXformEagerAgg.cpp]: remove the padding. we are slowly moving away from these decorations :)

File: libgpopt/src/xforms/CXformEagerAgg.cpp [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+    GPOS_ASSERT(NULL != pxf_ctxt);
+    GPOS_ASSERT(FPromising(pxf_ctxt->Pmp(), this, pexpr));
+    GPOS_ASSERT(FCheckPattern(pexpr));
+
+    IMemoryPool *mp = pxf_ctxt->Pmp();
+
+    // {
+    //  CAutoTrace at(mp);
+    //  at.Os() << *pexpr;
+    // }
```
Comment (vraghavan78) [inline libgpopt/src/xforms/CXformEagerAgg.cpp]: remove

File: libgpopt/src/xforms/CXformEagerAgg.cpp [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+    CColRefSet *outer_child_crs = CDrvdPropRelational::GetRelationalProperties(
+                                                        outer_child_expr->PdpDerive())->PcrsOutput();
+    CColRefSet *proj_list_crs = CDrvdPropScalar::GetDrvdScalarProps(
+                                                    proj_list_expr->PdpDerive())->PcrsUsed();
+    if (!outer_child_crs->ContainsAll(proj_list_crs))
+        return false;
+
+    /* 4. currently only supporting MIN aggregate function */
+    // obtain the oid of the scalar aggregate function and compare to the oid of
+    // EAggSum that applies to the input expression of the aggregate
```
Comment (vraghavan78) [inline libgpopt/src/xforms/CXformEagerAgg.cpp]: you mean MIN

File: libgpopt/src/xforms/CXformEagerAgg.cpp [hunk 7]
```diff
@@ -0,0 +1,390 @@
...
+    if (!agg_mdid->Equals(input_supported_agg_mdid))
+    {
+        // inputed agg is not supported
+        return false;
+    }
+
+    return true;
+}
+
+// populate the lower and upper aggreate's project list after
```
Comment (vraghavan78) [inline libgpopt/src/xforms/CXformEagerAgg.cpp]: aggreate's  -> aggregate's

File: libgpopt/src/xforms/CXformEagerAgg.cpp [hunk 8]
```diff
@@ -0,0 +1,390 @@
...
+
+    // list of project elements for the new lower and upper aggregates
+    CExpressionArray *lower_proj_elem_array = GPOS_NEW(mp) CExpressionArray(mp);
+    CExpressionArray *upper_proj_elem_array = GPOS_NEW(mp) CExpressionArray(mp);
+    const ULONG arity = orig_proj_list->Arity();
+
+    // loop over each project element
+    for (ULONG ul = 0; ul < arity; ul++)
+    {
+        CExpression *curr_proj_elem_expr = (*orig_proj_list)[ul];
```
Comment (vraghavan78) [inline libgpopt/src/xforms/CXformEagerAgg.cpp]: orig_proj_elem_expr

**Comment by iyerr3** [comment]
Replacing this with new PR using branch within this repo to allow collaborators edit the patch.

**Pull request closed without merging**